### PR TITLE
Use `IR.STRUCTURE_COMPARATOR` to compare two IR objects

### LINKED ISSUE #10985: Two IR objects aren't the same after de-serialization
```
 INFO ide_ci::program::command: sbt ℹ️ [info] java.io.IOException: Adding at 10443 object:
 INFO ide_ci::program::command: sbt ℹ️ [info]   org.enso.compiler.core.ir.Name$Literal@6d5a67af
 INFO ide_ci::program::command: sbt ℹ️ [info] but there already is:
 INFO ide_ci::program::command: sbt ℹ️ [info]   org.enso.compiler.core.ir.Name$Literal@5a63340d
 INFO ide_ci::program::command: sbt ℹ️ [info] are they equal: false
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:229)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:74)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:227)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:207)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.PersistName_Qualified.readObject(PersistName_Qualified.java:12)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.PersistName_Qualified.readObject(PersistName_Qualified.java:4)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.PersistImport_Module.readObject(PersistImport_Module.java:12)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.PersistImport_Module.readObject(PersistImport_Module.java:4)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.interpreter.caches.PersistBindingsMap_ResolvedImport.readObject(PersistBindingsMap_ResolvedImport.java:12)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.interpreter.caches.PersistBindingsMap_ResolvedImport.readObject(PersistBindingsMap_ResolvedImport.java:4)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readIndirect(PerInputImpl.java:218)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readObject(PerInputImpl.java:74)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:227)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrPersistance$PersistScalaList.readObject(IrPersistance.java:207)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.interpreter.caches.ImportExportCache$PersistBindingsMap.readObject(ImportExportCache.java:237)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.interpreter.caches.ImportExportCache$PersistBindingsMap.readObject(ImportExportCache.java:197)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance.readWith(Persistance.java:162)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:67)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.PerBufferReference.readObject(PerBufferReference.java:48)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.persist.Persistance$Reference.get(Persistance.java:253)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrLazyMap$En.getValue(IrLazyMap.java:57)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.IrLazyMap.get(IrLazyMap.java:37)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.ir.MetadataStorage.get(MetadataStorage.java:81)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.unsafeGetMetadata(Implicits.scala:108)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap$ResolvedModule.exportedSymbols(BindingsMap.scala:760)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap$ResolvedModule.findExportedSymbolsFor(BindingsMap.scala:752)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap$ResolvedImport.$anonfun$findExportedSymbolsFor$2(BindingsMap.scala:476)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/scala.collection.immutable.List.flatMap(List.scala:293)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap$ResolvedImport.findExportedSymbolsFor(BindingsMap.scala:476)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap.$anonfun$findExportedCandidatesInImports$1(BindingsMap.scala:150)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/scala.collection.immutable.List.flatMap(List.scala:293)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap.findExportedCandidatesInImports(BindingsMap.scala:147)
 INFO ide_ci::program::command: sbt ℹ️ [info] 	at org.enso.runtime/org.enso.compiler.data.BindingsMap.resolveName(BindingsMap.scala:184)
 ```
 (src https://github.com/enso-org/enso/actions/runs/10716335701/job/29713608555#step:7:1365)

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #10985 by providing a special `IR.STRUCTURE_COMPARATOR` that removes _metadata_ and _diagnostics_ before comparing two `IR` objects. Inefficient, not fully correct, but enough for the _debugging purposes_ this `prev != null` check was introduced.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Got the failure in [parallel PR CI run](https://github.com/enso-org/enso/actions/runs/11082351675/job/30795241816?pr=11207#step:7:1500). Need review!

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java [hunk 1]
```diff
@@ -28,6 +29,23 @@
...
+   * Compares IR structure, but not metadata neither diagnostics. A special comparator used by
+   * <em>Persistance API</em> to perform some rare consistency checks.
+   */
+  public static final Comparator<IR> STRUCTURE_COMPARATOR =
+      (aIr, bIr) -> {
+        if (aIr == bIr) {
+          return 0;
+        }
+        var aCopy = aIr.duplicate(true, false, false, true);
+        var bCopy = bIr.duplicate(true, false, false, true);
```
Comment (hubertp) [inline engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java:42]: That looks expensive if we do it for every IR that we read?

**Review by hubertp — APPROVED** [review]
Approving although I'm a bit wary of the performance impact. Please make sure it doesn't get worse.

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java:42]: We are not doing this for every `IR` we read. The code is only performed before exception:
```java
   sb.append("Adding at ").append(at).append(" object:\n  ");
      dumpObject(sb, res);
      sb.append("\nbut there already is:\n  ");
      dumpObject(sb, prev);
      sb.append("\nare they equal: ").append(bothObjectsAreTheSame);
```
is logged or thrown. How frequently that happens? It happens on CI from time to time. I haven't seen it on my computer yet.

**Pull request merged**